staging/src/k8s.io/client-go/tools/remotecommand/websocket.go

```diff
@@ -0,0 +1,342 @@
+	klog.V(4).Infof("The subprotocol is %s", streamingProto)
+	if streamingProto != remotecommand.StreamProtocolV5Name {
+		return fmt.Errorf("unsupported streaming protocol: %q", streamingProto)
+	}
+	streamer = newStreamProtocolV4(options)
+
+	creator := newWSStreamCreator(conn)
+	go creator.run()
+	return streamer.stream(creator)
+}
+
+type wsStreamCreator struct {
+	conn      *gwebsocket.Conn
+	connMu    sync.Mutex
+	streams   map[byte]*stream
+	streamsMu sync.Mutex
+}
+
+func newWSStreamCreator(conn *gwebsocket.Conn) *wsStreamCreator {
+	ws := wsStreamCreator{
+		conn:    conn,
+		streams: map[byte]*stream{},
+	}
+
+	go ws.sendPings(PingPeriod) // start heartbeat
+
+	return &ws
+}
+
+func (c *wsStreamCreator) getStream(id byte) *stream {
+	c.streamsMu.Lock()
+	defer c.streamsMu.Unlock()
+	return c.streams[id]
+}
+
+func (c *wsStreamCreator) setStream(id byte, s *stream) {
+	c.streamsMu.Lock()
+	defer c.streamsMu.Unlock()
+	c.streams[id] = s
+}
+
+// CreateStream uses id from passed headers to create a stream over "c.conn" connection.
+// Returns a Stream structure or nil and an error if one occurred.
+func (c *wsStreamCreator) CreateStream(headers http.Header) (httpstream.Stream, error) {
+	streamType := headers.Get(v1.StreamType)
+	id, ok := streamType2streamID[streamType]
+	if !ok {
+		return nil, fmt.Errorf("unknown stream type: %s", streamType)
+	}
+	if s := c.getStream(id); s != nil {
+		return nil, fmt.Errorf("duplicate stream for type %s", streamType)
+	}
+	reader, writer := io.Pipe()
+	s := &stream{
+		headers:   headers,
+		readPipe:  reader,
+		writePipe: writer,
+		conn:      c.conn,
+		connMu:    &c.connMu,
+		id:        id,
+	}
+	c.setStream(id, s)
+	return s, nil
+}
+
+// sendPings starts heartbeat sending a "ping" from the endpoint every "period".
+func (c *wsStreamCreator) sendPings(period time.Duration) {
+	c.connMu.Lock()
+	closedCh := make(chan struct{})
+	closeHandler := c.conn.CloseHandler()
+	c.conn.SetCloseHandler(func(code int, text string) error {
+		close(closedCh)
+		return closeHandler(code, text)
+	})
+	c.connMu.Unlock()
+	t := time.NewTicker(period)
+	defer t.Stop()
+	for {
+		select {
+		case <-closedCh:
+			return
+		case <-t.C:
+			c.connMu.Lock()
+			if err := c.conn.WriteControl(gwebsocket.PingMessage, nil, time.Now().Add(period)); err != nil {
+				klog.V(7).Infof("Websocket Ping failed: %v", err)
+				// Continue, in case this is a transient failure.
+				// c.conn.CloseChan above will tell us when the connection is
+				// actually closed.
+			} else {
+				klog.V(7).Infof("Websocket Ping succeeeded")
+			}
+			c.connMu.Unlock()
+		}
+	}
+}
+
+func (c *wsStreamCreator) run() {
+	readBuffer := make([]byte, remotecommand.WebSocketBufferSize)
+	// Set up handler for ping/pong heartbeat.
+	if err := c.conn.SetReadDeadline(time.Now().Add(pingReadDeadline)); err != nil {
+		klog.V(7).Infof("Websocket setting read deadline failed %v", err)
+	}
+	c.conn.SetPongHandler(func(string) error {
```

Comment:
is the server sending pings?

Reply:
Yes. I have manually verified the `golang.org/x/net/websockets` library automatically sends PONG in response to a PING:
```
I0713 14:37:50.169615 4081803 websocket.go:154] Starting Pings with 5000000000 period
I0713 14:37:50.169619 4081803 websocket.go:165] Websocket First Ping Sent...
I0713 14:37:50.169631 4081803 websocket.go:172] Websocket First Ping Sent...SUCCESSFUL
I0713 14:37:50.169733 4081803 websocket.go:205] Websocket Pong received...
```